app/src/main/java/com/fcaputo/parkingapp/mvp/model/ReservationModel.kt

```diff
@@ -0,0 +1,58 @@
+package com.fcaputo.parkingapp.mvp.model
+
+import com.fcaputo.parkingapp.entity.Reservation
+import com.fcaputo.parkingapp.mvp.contract.ReservationContract
+import com.fcaputo.parkingapp.utils.validation.ValidationErrorType
+import com.fcaputo.parkingapp.utils.validation.ValidationResult
+import java.time.LocalDateTime
+
+class ReservationModel : ReservationContract.Model {
+    private val parkingLotSize = 200 //TODO use the settings value
+    private val reservations: MutableList<Reservation> = mutableListOf()
+
+    override fun getParkingLotSize() = parkingLotSize
+
+    override fun storeReservation(reservation: Reservation) {
+        reservations.add(reservation)
+    }
+
+    override fun getReservations(): List<Reservation> = reservations.toList()
+
+    override fun validateData(startDate: LocalDateTime, endDate: LocalDateTime, parkingSpot: Int): ValidationResult {
+        return if (startDate.isBefore(LocalDateTime.now()) || endDate.isBefore(LocalDateTime.now())) {
+            ValidationResult(isSuccess = false, ValidationErrorType.DATETIMES_ARE_PAST)
+        } else if (startDate.isAfter(endDate)) {
+            ValidationResult(isSuccess = false, ValidationErrorType.DATETIMES_ARE_UNORDERED)
+        } else if (parkingSpot == 0) {
+            ValidationResult(isSuccess = false, ValidationErrorType.SPOT_IS_ZERO)
+        } else if (parkingSpot > parkingLotSize) {
+            ValidationResult(isSuccess = false, ValidationErrorType.SPOT_IS_OUT_OF_RANGE)
+        } else if (spotIsUnavailable(startDate, endDate, getReservationsByParkingSpot(parkingSpot))) {
+            ValidationResult(isSuccess = false, ValidationErrorType.SPOT_IS_UNAVAILABLE)
+        } else {
+            ValidationResult(isSuccess = true)
+        }
+    }
+
+    private fun getReservationsByParkingSpot(spotNumber: Int) = reservations.filter{ it.parkingLot == spotNumber }.toList()
+
+    private fun spotIsUnavailable(startDate: LocalDateTime, endDate: LocalDateTime, spotReservations: List<Reservation>): Boolean {
+        var isUnavailable = false
+        for (res in spotReservations) {
+            if (datesOverlap(startDate, endDate, res)) {
+                isUnavailable = true
+                break
+            }
+        }
+
+        return isUnavailable
```

Comment:
Minor improvement: you could use the extension function `spotReservations.any { }` on collections!